lib/ra10ke/validate.rb

```diff
@@ -0,0 +1,127 @@
+# frozen_string_literal: true
+
+require 'ra10ke/monkey_patches'
+require 'tempfile'
+require 'table_print'
+require 'ra10ke/puppetfile_parser'
+require 'English'
+
+module Ra10ke
+  module Validate
+
+    GOOD_EMOJI = ENV['GOOD_EMOJI'] || '👍'
+    BAD_EMOJI = ENV['BAD_EMOJI'] || '😨'
+
+    # Validate the git urls and refs
+    def define_task_validate(*args)
+      desc 'Validate the git urls and branches, refs, or tags'
+      task :validate do
+        gitvalididation = Ra10ke::Validate::Validation.new(get_puppetfile.puppetfile_path)
+        exit_code = 0
+        if gitvalididation.bad_mods?
+          exit_code = 1
+          message = BAD_EMOJI + '  Not all modules in the Puppetfile are valid. '.red + BAD_EMOJI
+        else
+          message = GOOD_EMOJI + '  Puppetfile looks good. '.green + GOOD_EMOJI
+        end
+        tp gitvalididation.sorted_mods, :name, { url: { width: 50 } }, :ref, :status
+        abort(message) if exit_code.positive?
+        puts message
+      end
+    end
+
+    class Validation
+      include Ra10ke::PuppetfileParser
+      
+      attr_reader :puppetfile
+
+      def initialize(file)
+        file ||= './Puppetfile'
+        @puppetfile = File.expand_path(file)
+        abort("Puppetfile does not exist at #{puppetfile}") unless File.readable?(puppetfile)
+      end
+
+      # @return [Boolean] - return true if the ref is valid
+      # @param url [String] - the git string either https or ssh url
+      # @param ref [String] - the ref object, branch name, tag name, or commit sha, defaults to HEAD
+      def valid_ref?(url, ref = 'HEAD')
+        raise ArgumentError unless ref
+        found = all_refs(url).find do |sha, data | 
+          # we don't need to bother with these types
+          next if data[:type] == :pull || data[:type] == :merge_request     
+          # is the name equal to the tag or branch?  Is the commit sha equal?     
+          data[:name].eql?(ref) || sha.slice(0,8).eql?(ref.slice(0,8))
+        end
+        !found.nil?
+      end
+
+      # @return [Hash] - a hash of all the refs associated with the remote repository
+      # @param url [String] - the git string either https or ssh url
+      # @example
+      # {"0ec707e431367bbe2752966be8ab915b6f0da754"=>{:ref=>"refs/heads/74110ac", :type=>:branch, :subtype=>nil, :name=>"74110ac"},
+        # "07bb5d2d94db222dca5860eb29c184e8970f36f4"=>{:ref=>"refs/pull/74/head", :type=>:pull, :subtype=>:head, :name=>"74"},
+        # "156ca9a8ea69e056e86355b27d944e59d1b3a1e1"=>{:ref=>"refs/heads/master", :type=>:branch, :subtype=>nil, :name=>"master"},
+        # "fcc0532bbc5a5b65f3941738339e9cc7e3d767ce"=>{:ref=>"refs/pull/249/head", :type=>:pull, :subtype=>:head, :name=>"249"},
+        # "8d54891fa5df75890ee15d53080c2a81b4960f92"=>{:ref=>"refs/pull/267/head", :type=>:pull, :subtype=>:head, :name=>"267"} }
+      def all_refs(url)
```

Comment:
@ZeroPointEnergy This is the new method I added to fully inspect all the refs.  Grep was not enough as it turns out.  valid_ref? was refactored as well.

Reply:
Very nice!! :+1: